flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/BufferManager.java

```diff
@@ -151,23 +153,63 @@ int requestFloatingBuffers(int numRequired) throws IOException {
 				return numRequestedBuffers;
 			}
 
-			numRequiredBuffers = numRequired;
+			numRequiredBuffers += backlog;
+			numRequestedBuffers = internalRequestFloatingBuffers(numRequiredBuffers);
+			numRequiredBuffers -= numRequestedBuffers;
+		}
+		return numRequestedBuffers;
+	}
 
-			while (bufferQueue.getAvailableBufferSize() < numRequiredBuffers && !isWaitingForFloatingBuffers) {
-				BufferPool bufferPool = inputChannel.inputGate.getBufferPool();
-				Buffer buffer = bufferPool.requestBuffer();
-				if (buffer != null) {
-					bufferQueue.addFloatingBuffer(buffer);
-					numRequestedBuffers++;
-				} else if (bufferPool.addBufferListener(this)) {
-					isWaitingForFloatingBuffers = true;
-					break;
-				}
+	private int internalRequestFloatingBuffers(int numBuffersToRequest) throws IOException {
+		assert Thread.holdsLock(bufferQueue);
+
+		int numRequestedBuffers = 0;
+		while (numRequestedBuffers < numBuffersToRequest && !isWaitingForFloatingBuffers) {
+			BufferPool bufferPool = inputChannel.inputGate.getBufferPool();
+			Buffer buffer = bufferPool.requestBuffer();
+			if (buffer != null) {
+				bufferQueue.addFloatingBuffer(buffer);
+				numRequestedBuffers++;
+			} else if (bufferPool.addBufferListener(this)) {
+				isWaitingForFloatingBuffers = true;
+				break;
 			}
 		}
 		return numRequestedBuffers;
 	}
 
+	public NettyMessage.ResumeConsumption resumeAndGetResumptionMessage(
+			InputChannelID channelID,
+			int initialCredit) throws IOException {
+		synchronized (bufferQueue) {
+			checkState(numRequiredBuffers >= 0, "Number of required buffers should be non-negative.");
+			checkState(bufferQueue.getAvailableBufferSize() == initialCredit, "Illegal number of available buffers.");
+
+			if (initialCredit > 0) {
+				return new NettyMessage.ResumeConsumption(channelID, initialCredit, numRequiredBuffers);
+			}
+
+			int numCredit = internalRequestFloatingBuffers(numRequiredBuffers);
+			numRequiredBuffers -= numCredit;
+			return new NettyMessage.ResumeConsumption(channelID, numCredit, numRequiredBuffers);
+		}
+	}
+
+	public void onCheckpointBarrier(CheckpointBarrier barrier, int initialCredit) {
```

Comment:
Keep in mind that `BufferManager` is only for buffer request/release related operations, so it should not understand the other specific logics which should be done inside the respective `InputChannel`. Otherwise we would dirty this component and have ambiguous definition what is the role of this component.

In detail, the following should be done inside `RemoteInputChannel`

```
CheckpointOptions options = barrier.getCheckpointOptions();
if (initialCredit == 0 && options.isExactlyOnceMode() && !options.isUnalignedCheckpoint()) 
```
And rename  the method `onCheckpointBarrier` to distinguish with existing `releaseFloatingBuffers()`. From the perspective of outside caller, we should give a clear semantic method naming in order to be reused future.